### PR TITLE
dd 13 stations management

### PR DESCRIPTION
**Pull request by ChickenMcNuggetss**
# Task

[Stations management](https://divas801801.atlassian.net/jira/software/projects/DD/boards/2?selectedIssue=DD-13)

# Description

- Station coordinates input;
- Connecting stations;
- Displaying existing stations;
- Deleting stations;

## Pull request type

Please check the type of change your PR introduces:

- [x] 🍕 Feature
- [ ] 🐛 Bugfix
- [ ] 🧑‍💻 Code Refactor
- [ ] 📝 Documentation Update
- [ ] 🤖 Build related changes
- [ ] 🔥 Performance Improvements
- [ ] Other (please describe):

# Demo (situationally)

# Has It Been Tested?

- [ ] Yes
- [ ] No, and this is why: please replace this line with details on why tests have not been included
- [ ] Tests are not needed for this task
- [ ] I need help with writing tests

### CODE REVIEW

File: src/app/features/admin/validators/longitude.validator.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';
+
+export function longitudeValidator(): ValidatorFn {
+  return (control: AbstractControl): ValidationErrors | null => {
+    let isLongitudeValid;
+    if (control.value <= 180 && control.value >= -180) isLongitudeValid = true;
```
Comment (TELEUZI) [inline src/app/features/admin/validators/longitude.validator.ts]: Magic number.

File: src/app/features/admin/validators/latitude.validator.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';
+
+export function latitudeValidator(): ValidatorFn {
+  return (control: AbstractControl): ValidationErrors | null => {
+    let isLatitudeValid;
+    if (control.value <= 90 && control.value >= -90) isLatitudeValid = true;
```
Comment (TELEUZI) [inline src/app/features/admin/validators/latitude.validator.ts]: Magic number.

File: src/app/features/admin/validators/longitude.validator.ts (hunk 1 above)
Comment (TELEUZI) [inline src/app/features/admin/validators/longitude.validator.ts]: Use ternary and `const`

File: src/app/features/admin/validators/latitude.validator.ts (hunk 2 above)
Comment (TELEUZI) [inline src/app/features/admin/validators/latitude.validator.ts]: Use ternary and `const`

File: src/app/features/admin/pages/stations-page/stations-page.component.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  private readonly destroy = inject(DestroyRef);
+
+  public ngOnInit() {
+    this.stationsApiService
+      .retrieveStationList()
+      .pipe(takeUntilDestroyed(this.destroy))
+      .subscribe((stations) => {
+        this.stationsService.setStations(stations);
+      });
+  }
```
Comment (TELEUZI) [inline src/app/features/admin/pages/stations-page/stations-page.component.ts]: Why use `ngOnInit`?

File: src/app/features/admin/pages/stations-page/stations-page.component.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  private readonly stationsService = inject(StationsService);
+
+  private readonly destroy = inject(DestroyRef);
+
+  public ngOnInit() {
+    this.stationsApiService
+      .retrieveStationList()
+      .pipe(takeUntilDestroyed(this.destroy))
+      .subscribe((stations) => {
+        this.stationsService.setStations(stations);
```
Comment (TELEUZI) [inline src/app/features/admin/pages/stations-page/stations-page.component.ts]: Loader?

File: src/app/features/admin/components/existing-stations-list/existing-stations-list.component.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+  public readonly stations = this.stationsService.stations;
+
+  public deleteStation(id: number) {
+    let hasActiveRide;
+    this.stationsApiService
+      .retrieveOrders()
+      .pipe(takeUntilDestroyed(this.destroy))
+      .subscribe((orders) => {
+        hasActiveRide = orders.some((order) => order.status === 'active' && order.path.includes(id));
```
Comment (TELEUZI) [inline src/app/features/admin/components/existing-stations-list/existing-stations-list.component.ts]: How this should work if code with `if (hasActiveRide)` would be executed earlier then `subscribe`'s callback?

File: src/app/features/admin/components/existing-stations-list/existing-stations-list.component.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      });
+    if (hasActiveRide) {
+      this.alert.open({ message: 'Cannot delete station with active rides', label: 'Error', appearance: 'error' });
+      return;
+    }
+    this.stationsApiService.deleteStation(id);
+    this.stationsService.deleteStationFromList(id);
```
Comment (TELEUZI) [inline src/app/features/admin/components/existing-stations-list/existing-stations-list.component.ts]: Are you sure that `deleteStation` would always work and we always should execute`deleteStationFromList`?

File: src/app/features/admin/components/existing-stations-list/existing-stations-list.component.html [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    </header>
+
+    <section>
+      <div>
+        <div><span class="tui-text_body-l">Latitude: </span>{{ station.latitude }}</div>
+        <div><span class="tui-text_body-l">Longitude: </span>{{ station.longitude }}</div>
+      </div>
+      <span class="tui-text_body-l">Connected to:</span>
+      <span *ngFor="let station of station.connectedTo"> {{ getStations(station.id)?.city }} </span>
+    </section>
```
Comment (TELEUZI) [inline src/app/features/admin/components/existing-stations-list/existing-stations-list.component.html]: Why not just fetch all stations at once and don't `find` on every station?

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+  public ngOnInit() {
+    this.connectedStations.valueChanges.pipe(debounceTime(1000), takeUntilDestroyed(this.destroy)).subscribe(() => {
+      const lastControl = this.connectedStations.at(this.connectedStations.length - 1);
+      if (lastControl.value !== '') {
+        this.connectedStations.push(this.createConnectedStationControl());
+      }
+    });
+
+    this.cityName.valueChanges.pipe(debounceTime(1000), takeUntilDestroyed(this.destroy)).subscribe((city) => {
+      if (!city) return;
```
Comment (TELEUZI) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: Use `filter` operator

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+      this.locationService
+        .getLocationCoordinates(city)
```
Comment (TELEUZI) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: Use `switchMap` operator, there should be no `subscribe` in `subscribe`

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  public createStation() {
+    if (!this.createStationForm.valid) return;
+    if (this.stationsService.stations().find((station) => station.city === this.cityName.value)) return;
+    if (!this.cityName.value || !this.latitude.value || !this.longitude.value) return;
+    const body: NewStation = {
+      city: this.cityName.value,
+      latitude: +this.latitude.value,
+      longitude: +this.longitude.value,
```
Comment (TELEUZI) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: Why not just create `number` inputs/controls, not `string` inputs?

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+  public onSelected(station: StationListItem, index: number) {
+    if (this.relations.includes(index)) return;
+    this.connectedStations.controls[index].setValue(station.city);
+    this.relations[index] = station.id;
+  }
+
+  public createStation() {
+    if (!this.createStationForm.valid) return;
+    if (this.stationsService.stations().find((station) => station.city === this.cityName.value)) return;
+    if (!this.cityName.value || !this.latitude.value || !this.longitude.value) return;
```
Comment (TELEUZI) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: Validators?

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+      latitude: +this.latitude.value,
+      longitude: +this.longitude.value,
+      relations: this.relations,
+    };
+    this.stationsApiService
+      .createNewStation(body)
+      .pipe(takeUntilDestroyed(this.destroy))
+      .subscribe({
+        next: ({ id }) => {
+          this.createdStationId = id;
```
Comment (TELEUZI) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts:151]: Use `tap` operator for side-effects.

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+          this.stationsApiService
+            .retrieveStationList()
+            .pipe(takeUntilDestroyed(this.destroy))
```
Comment (TELEUZI) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: Use `switchMap` operator

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    return this.createStationForm.controls.cityName;
+  }
+
+  public get latitude() {
+    return this.createStationForm.controls.latitude;
+  }
+
+  public get longitude() {
+    return this.createStationForm.controls.longitude;
+  }
```
Comment (TELEUZI) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: Just create `controls` getter?

File: src/app/features/admin/components/existing-stations-list/existing-stations-list.component.html [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+      <div tuiTitle="m">{{ station.city }}</div>
+      <button
+        appearance="secondary"
+        size="m"
+        tuiButton
+        type="button"
+        (click)="deleteStation(station.id)"
+        class="tui-form__button"
+      >
+        <img alt="" [src]="'@tui.trash' | tuiIcon" />
```
Comment (TELEUZI) [inline src/app/features/admin/components/existing-stations-list/existing-stations-list.component.html]: Empty `alt`

File: src/app/features/admin/components/existing-stations-list/existing-stations-list.component.html [hunk 16]
```diff
@@ -0,0 +1,26 @@
+<div tuiPlatform="web" *ngFor="let station of stations()">
```
Comment (TELEUZI) [inline src/app/features/admin/components/existing-stations-list/existing-stations-list.component.html]: trackBy

File: src/app/features/admin/components/existing-stations-list/existing-stations-list.component.html [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+      </button>
+    </header>
+
+    <section>
+      <div>
+        <div><span class="tui-text_body-l">Latitude: </span>{{ station.latitude }}</div>
+        <div><span class="tui-text_body-l">Longitude: </span>{{ station.longitude }}</div>
+      </div>
+      <span class="tui-text_body-l">Connected to:</span>
+      <span *ngFor="let station of station.connectedTo"> {{ getStations(station.id)?.city }} </span>
```
Comment (TELEUZI) [inline src/app/features/admin/components/existing-stations-list/existing-stations-list.component.html]: trackBy

File: src/app/features/admin/components/create-station-form/create-station-form.component.html [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    formArrayName="connectedStations"
+    *ngFor="let tag of connectedStations.controls; let i = index"
+    class="tui-form__multi-fields"
+  >
+    <tui-input *tuiLet="stations() as stations" [formControlName]="i" class="tui-form__row tui-space_vertical-4">
+      Connected City
+      <input placeholder="Type city name" tuiTextfieldLegacy />
+      <ng-container *ngIf="stations?.length">
+        <tui-data-list *tuiDataList>
+          <button *ngFor="let station of stations" tuiOption [value]="station.city" (click)="onSelected(station, i)">
```
Comment (TELEUZI) [inline src/app/features/admin/components/create-station-form/create-station-form.component.html]: trackBy

File: src/app/features/admin/components/create-station-form/create-station-form.component.html [hunk 19]
```diff
@@ -0,0 +1,51 @@
+<div class="tui-form__header">Create station</div>
+<form class="tui-form__multi-fields" [formGroup]="createStationForm" (ngSubmit)="createStation()">
+  <tui-input *tuiLet="cities() as cities" [formControl]="cityName" class="tui-form__row">
+    City Name
+    <input placeholder="Type city name" tuiTextfieldLegacy />
+    <ng-container *ngIf="cities?.length">
+      <tui-data-list *tuiDataList="">
+        <button *ngFor="let city of cities" tuiOption [value]="city.name" (click)="onSelectedCity(city)">
```
Comment (TELEUZI) [inline src/app/features/admin/components/create-station-form/create-station-form.component.html]: trackBy

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+    latitude: this.fb.control('', [Validators.required, latitudeValidator()]),
+    longitude: this.fb.control('', [Validators.required, longitudeValidator()]),
+    connectedStations: this.fb.array([this.createConnectedStationControl()]),
+  });
+
+  // TODO: place the retrieveStationList method in another location in code
+  public ngOnInit() {
+    this.connectedStations.valueChanges.pipe(debounceTime(1000), takeUntilDestroyed(this.destroy)).subscribe(() => {
+      const lastControl = this.connectedStations.at(this.connectedStations.length - 1);
+      if (lastControl.value !== '') {
```
Comment (TELEUZI) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: Use `filter` operator.

File: src/app/features/admin/pages/stations-page/stations-page.component.ts (hunk 3 above)
Comment (TELEUZI) [reply src/app/features/admin/pages/stations-page/stations-page.component.ts]: Check using angular `resolvers`

File: src/app/features/admin/components/existing-stations-list/existing-stations-list.component.ts (hunk 6 above)
Comment (ChickenMcNuggetss) [reply src/app/features/admin/components/existing-stations-list/existing-stations-list.component.ts:43]: Yup, it should

File: src/app/features/admin/validators/longitude.validator.ts (hunk 1 above)
Comment (ChickenMcNuggetss) [reply src/app/features/admin/validators/longitude.validator.ts]: You know, I'm something of a wizard myself
image

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+import { AlertService } from '@core/services/alert/alert.service';
+import { NewStation } from '@features/admin/interfaces/new-station.interface';
+import { StationListItem } from '@features/admin/interfaces/station-list-item.interface';
+import { StationsApiService } from '@features/admin/services/stations-api/stations-api.service';
+import { StationsService } from '@features/admin/services/stations/stations.service';
+import { latitudeValidator } from '@features/admin/validators/latitude.validator';
+import { longitudeValidator } from '@features/admin/validators/longitude.validator';
+import { CityInfo } from '@features/home/interfaces/city-info.interface';
+import { LocationApiService } from '@features/home/services/location-api/location-api.service';
+import { SearchService } from '@features/home/services/search/search.service';
```
Comment (Dewlar) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: incorrect import lint: must be from `search`

File: src/app/features/admin/components/existing-stations-list/existing-stations-list.component.ts [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+      .retrieveOrders()
+      .pipe(takeUntilDestroyed(this.destroy))
+      .subscribe((orders) => {
+        hasActiveRide = orders.some((order) => order.status === 'active' && order.path.includes(id));
+      });
+    if (hasActiveRide) {
+      this.alert.open({ message: 'Cannot delete station with active rides', label: 'Error', appearance: 'error' });
+      return;
+    }
+    this.stationsApiService.deleteStation(id);
```
Comment (Dewlar) [inline src/app/features/admin/components/existing-stations-list/existing-stations-list.component.ts]: no subscription - no request

**Review by Dewlar — DISMISSED** [review]

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts (hunk 21 above)
Comment (ChickenMcNuggetss) [reply src/app/features/admin/components/create-station-form/create-station-form.component.ts]: Yep, I already fixed that, thank you

File: src/app/features/admin/components/create-station-form/create-station-form.component.html [hunk 23]
```diff
@@ -0,0 +1,70 @@
+<div class="tui-form__header">Create station</div>
+<form class="tui-form__multi-fields" [formGroup]="createStationForm" (ngSubmit)="createStation()">
+  <div class="container">
+    <div class="input-container">
+      <tui-input *tuiLet="cities() as cities" [formControl]="cityName" class="tui-form__row input-container">
+        City Name
+        <input placeholder="Type city name" tuiTextfieldLegacy />
+        <ng-container *ngIf="cities?.length">
```
Comment (RykovaMariia) [inline src/app/features/admin/components/create-station-form/create-station-form.component.html]: Do we need "?" here?

File: src/app/features/admin/components/create-station-form/create-station-form.component.html [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+          <tui-data-list *tuiDataList="">
```
Comment (RykovaMariia) [inline src/app/features/admin/components/create-station-form/create-station-form.component.html]:
```suggestion
          <tui-data-list *tuiDataList>
```

File: src/app/features/admin/components/create-station-form/create-station-form.component.html [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+        </ng-container>
+      </tui-input>
+      <ng-container
+        [ngTemplateOutlet]="input"
+        [ngTemplateOutletContext]="{ label: 'Latitude', control: 'latitude' }"
+      ></ng-container>
+      <ng-container
+        [ngTemplateOutlet]="input"
+        [ngTemplateOutletContext]="{ label: 'Longitude', control: 'longitude' }"
+      ></ng-container>
```
Comment (RykovaMariia) [inline src/app/features/admin/components/create-station-form/create-station-form.component.html]: maybe use` | titlecase`

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+export class CreateStationFormComponent implements OnInit {
+  private readonly searchService = inject(SearchService);
+
+  private readonly stationsApiService = inject(StationsApiService);
+
+  private readonly stationsService = inject(StationsService);
+
+  private readonly locationService = inject(LocationApiService);
+
+  public cities = this.searchService.cities;
```
Comment (RykovaMariia) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: https://discord.com/channels/1272883598915014676/1276198159768621126/1276202491134087200

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+    longitude: this.fb.control<number | null>(null, [Validators.required, longitudeValidator()]),
+    connectedStations: this.fb.array([this.fb.control('', Validators.required)]),
+  });
+
+  // TODO: place the retrieveStationList method in another location in code
+  public ngOnInit() {
+    this.connectedStations.valueChanges
+      .pipe(
+        debounceTime(1000),
+        filter(() => this.connectedStations.at(this.connectedStations.length - 1).value !== ''),
```
Comment (RykovaMariia) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: at(-1) ?

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+      )
+      .subscribe(() => {
+        this.connectedStations.controls[this.connectedStations.length - 1].addValidators(Validators.required);
+        this.connectedStations.push(this.createConnectedStationControl());
+      });
+
+    this.cityName.valueChanges
+      .pipe(
+        debounceTime(1000),
+        filter((city) => city !== null),
```
Comment (RykovaMariia) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: filter((city) => !city)

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 29]
```diff
@@ -0,0 +1,156 @@
...
+    };
+    this.stationsApiService
+      .createNewStation(body)
+      .pipe(
+        tap(({ id }) => {
+          this.createdStationId = id;
+        }),
+        switchMap(() => this.stationsApiService.retrieveStationList()),
+        map((stations) => stations.find((station) => station.id === this.createdStationId)),
+        filter((station) => station !== undefined),
```
Comment (RykovaMariia) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]:
```suggestion
         filter((station) => !station)
```

File: src/app/features/admin/resolvers/stations-resolver.ts [hunk 30]
```diff
@@ -0,0 +1,20 @@
...
+  const stationsApiService = inject(StationsApiService);
+  const stationsService = inject(StationsService);
+
+  switch (state.url) {
+    case '/admin/stations':
+      return stationsApiService.retrieveStationList().pipe(tap((stations) => stationsService.setStations(stations)));
+    default:
+      return EMPTY;
+  }
+};
```
Comment (Dewlar) [inline src/app/features/admin/resolvers/stations-resolver.ts]: ternary ?

File: src/app/features/admin/components/create-station-form/create-station-form.component.ts [hunk 31]
```diff
@@ -0,0 +1,156 @@
...
+    return this.controls.cityName;
+  }
+
+  public get latitude() {
+    return this.controls.latitude;
+  }
+
+  public get longitude() {
+    return this.controls.longitude;
+  }
```
Comment (RykovaMariia) [inline src/app/features/admin/components/create-station-form/create-station-form.component.ts]: delete, please

File: src/app/features/admin/validators/latitude.validator.ts [hunk 32]
```diff
@@ -0,0 +1,11 @@
+import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';
+
+const latitudeBoundaryValue = 90;
```
Comment (Dewlar) [inline src/app/features/admin/validators/latitude.validator.ts:3]: magic number (c) Yauheni

File: src/app/features/admin/validators/longitude.validator.ts [hunk 33]
```diff
@@ -0,0 +1,11 @@
+import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';
+
+const longitudeBoundaryValue = 180;
```
Comment (Dewlar) [inline src/app/features/admin/validators/longitude.validator.ts:3]: magic number (c) Yauheni

File: src/app/features/admin/components/existing-stations-list/existing-stations-list.component.ts [hunk 34]
```diff
@@ -0,0 +1,61 @@
...
+
+  private readonly alert = inject(AlertService);
+
+  public readonly stations = this.stationsService.stations;
+
+  public deleteStation(id: number) {
+    this.stationsApiService
+      .retrieveOrders()
+      .pipe(takeUntilDestroyed(this.destroy))
+      .subscribe((orders) => {
```
Comment (RykovaMariia) [inline src/app/features/admin/components/existing-stations-list/existing-stations-list.component.ts]: please add error: {}

File: src/app/features/admin/validators/latitude.validator.ts (hunk 32 above)
Comment (ChickenMcNuggetss) [inline src/app/features/admin/validators/latitude.validator.ts:3]: How to fix?

File: src/app/features/admin/validators/longitude.validator.ts (hunk 33 above)
Comment (ChickenMcNuggetss) [inline src/app/features/admin/validators/longitude.validator.ts:3]: How to fix? (c) ChickenMcNuggetss

File: src/app/features/search/services/location-api/location-api.service.ts [hunk 35]
```diff
@@ -16,4 +16,10 @@ export class LocationApiService {
       `${environment.baseLocationApiUrl}/direct?q=${city}&limit=${responseLimit}&appid=${environment.locationApiKey}`
     );
   }
+
+  public getLocation(lat: number, lon: number) {
+    return this.httpClient.get(
+      `${environment.baseLocationApiUrl}/reverse?lat=${lat}&lon=${lon}&limit=1&appid=${environment.locationApiKey}`
+    );
+  }
```
Comment (RykovaMariia) [inline src/app/features/search/services/location-api/location-api.service.ts]: it is not used anywhere

**Review by RykovaMariia — APPROVED** [review]

**Review by Dewlar — APPROVED** [review]

**Pull request merged**